### PR TITLE
Move version_counter_ to TensorImpl

### PR DESCRIPTION
**Pull request by yf225**
According to https://github.com/pytorch/pytorch/issues/13638#issuecomment-468055428, after the Variable/Tensor merge, we may capture variables without autograd metadata inside an autograd function, and we need a working version counter in these cases. This PR makes it possible by moving `version_counter_` out of autograd metadata and into TensorImpl, so that variables without autograd metadata still have version counters.

### CODE REVIEW

**Comment by ezyang** [comment]
Hmm. I expected the version counter to go on storage, because you care about mutations any time you touch the storage, not just the tensor itself. The old Variable code had logic in view to make sure the shared pointer gets shared across all variable views, but if you have a non-autograd tensor, it won't hit that codebase and the sharing will not be setup appropriately.  This is not observable today because you haven't flipped the switch yet.

Once you put it on storage you can get rid of the `shared_ptr`; it was only needed because of this aliasing behavior.

**Review by ezyang — CHANGES_REQUESTED** [review]
Doesn't look right

**Comment by gchanan** [comment]
@ezyang you can't put the version counter on Storage because `detach`, which is usable on sparse tensors, also shares the version counter, but sparse tensors don't have Storage.

I'm not sure exactly which model we are using here for "flip the switch."

**Comment by ezyang** [comment]
What I mean by flip the switch, is capture a non-Variable Tensor in SavedVariables. You won't have accurate tracking in that case.

If there isn't a way around the sparse tensor problem, then we need to audit all of the non-Variable view operations and make sure they preserve the version counter appropriately as well.

**Comment by yf225** [comment]
Based on offline chat with @ezyang : 
1. We should put the `version_counter_` increment in non-Variable dispatch paths as well, so that `version_counter_` accurately reflects the update count of this tensor. 
2. We need to make sure that `shallow_copy_and_detach` copies the `version_counter_` shared pointer, so that any change to the original tensor's storage is reflected in the shallow copy.

**Comment by yf225** [comment]
@ezyang @gchanan This PR is ready for review again. Thx :)

**Comment by yf225** [comment]
I updated the PR description to show the detailed changes involved.

**Comment by yf225** [comment]
NOTE: add example to demonstrate why this is needed.

File: test/cpp/api/misc.cpp
Comment (yf225) [inline test/cpp/api/misc.cpp]: @gchanan These two are the use cases we care about, and they will still work after Variable/Tensor merge. Detailed discussion is at the end of the PR description.

File: aten/src/ATen/SparseTensorImpl.h
Comment (ezyang) [inline aten/src/ATen/SparseTensorImpl.h]: Err, really? I thought if you do a shallow copy, the version counter should always come along for the ride, because the storage is being shared. Looking at the note, I agree that `set_data` should not share, but that seems like the odd case out of all the rest.

File: aten/src/ATen/core/Tensor.h
Comment (ezyang) [inline aten/src/ATen/core/Tensor.h]: Seeing these on the public API makes me a little unhappy, because we don't have them on `torch.Tensor`, but I guess they were on `Variable` before so this is the easiest thing to do for now.

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: Nice!

File: c10/core/TensorImpl.h
Comment (ezyang) [inline c10/core/TensorImpl.h]: You should put two entries here, because it costs two words.

File: test/test_cuda.py
Comment (ezyang) [inline test/test_cuda.py]: What happened here?

File: test/test_nn.py
Comment (ezyang) [inline test/test_nn.py]: This is a nice catch.

However, assuming that what we are trying to determine is that `input_ip_clone` was in-place mutated, wouldn't it be easier to just compare it against what its version was before we called `module_ip` on it? (Or is the test trying to assert that the version counters are not shared? That's easy enough to test too: additionally check that input version didn't bump.)

File: tools/autograd/gen_variable_type.py
Comment (ezyang) [inline tools/autograd/gen_variable_type.py]: This assert is very nice, thank you!

File: torch/csrc/autograd/variable.cpp
Comment (ezyang) [inline torch/csrc/autograd/variable.cpp]: I'm confused by this. I don't see any mutating operations, so I wouldn't expect that the version counter wouldn't get updated in this case. Are you trying to prevent sharing from taking place? But surely `set_version_counter` isn't the appropriate way to go about doing this?

**Comment by ezyang** [comment]
Holding off on accept because I am legitimately perplexed about the `set_data` implementation and what's going on with `shallow_copy_and_detach`.

**Comment by ezyang** [comment]
Also, which part of the code is responsible for making sure a view of a tensor shares version counter with its base?

File: test/test_cuda.py
Comment (yf225) [reply test/test_cuda.py]: After moving version bumping to non-Variable dispatch functions, there is no guarantee that the version will be bumped only once in an in-place operation. In this case, `t.zero_()` bumps `t`'s version twice, first in:
```cpp
Tensor & TypeDefault::zero_(Tensor & self) const {
    const OptionalDeviceGuard device_guard(device_of(self));
    if (at::VersionUpdateMode::is_enabled()) self.bump_version();  // NOTE: bumps version
    // self.bump_version();
    return at::native::zero_(/* native_actuals */ self);
}
```
and then in:
```cpp
Tensor & CPUFloatType::_th_zero_(Tensor & self) const {
    // DeviceGuard omitted
    if (at::VersionUpdateMode::is_enabled()) self.bump_version();  // NOTE: bumps version
    // self.bump_version();
    auto self_ = checked_tensor_unwrap(self,"self",1, false, Backend::CPU, ScalarType::Float);
    THFloatTensor_zero(self_);
    return self;
}
```

File: torch/csrc/autograd/variable.h
Comment (gchanan) [inline torch/csrc/autograd/variable.h]: curious: can you not put override on the implementation below?

Comment (yf225) [reply torch/csrc/autograd/variable.h]: Just tried and yes we can :)

**Comment by yf225** [comment]
@ezyang These two places are responsible for making sure a view of a tensor shares version counter with its base:
1.  https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/variable.h#L569
2. https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/variable.cpp#L196

We don't need to change their logic in this PR, because they will still work properly.

File: torch/csrc/autograd/variable.cpp
Comment (yf225) [reply torch/csrc/autograd/variable.cpp]: Fixed

File: torch/autograd/gradcheck.py
Comment (yf225) [inline torch/autograd/gradcheck.py]: I would like to propose the `with torch.no_grad(update_version=False)` API, because this function (`get_numerical_jacobian(...)`, which is a crucial function used by `gradcheck()` and `gradgradcheck()`) wants the original version counter values of `x_tensor` to always be preserved, so that making small finite changes to `x_tensor` (and restoring its original value later) doesn't invalidate it in the autograd graph.

File: aten/src/ATen/version_update_mode.h
Comment (ezyang) [inline aten/src/ATen/version_update_mode.h]: This needs a comment

File: torch/autograd/gradcheck.py
Comment (ezyang) [reply torch/autograd/gradcheck.py]: OK, this looks sound to me.

I might explain what's going on this code a little differently. Basically, the invariant required by version is that we must update the version if the contents of the tensor change. We are preserving this invariant because we reset the tensor to its original value after running the finite differences, but PyTorch doesn't know that so we help it out.

Your comment says that it's important not to update the version counter, but it doesn't explain why *not* updating the version counter is sound. That's an important thing to write down.

File: aten/src/ATen/version_update_mode.h
Comment (ezyang) [reply aten/src/ATen/version_update_mode.h]: There are some places in code where it is not sound to toggle the enabled setting. We should explain when this is the case. It's something like, you're not allowed to set this when you're "inside" a Variable operation.

**Review by ezyang — APPROVED** [review]
I still suspect that we are kicking some view and version counter related needed diffs down the road, but this is OK to go in for now.

**Comment by gchanan** [comment]
if `update_version` is only used for gradcheck why do we need to expose it to users?

File: aten/src/ATen/SparseTensorImpl.h
Comment (yf225) [inline aten/src/ATen/SparseTensorImpl.h]: This is a missing check for this function: we don't want users to call `shallow_copy_and_detach()` on a Variable, because Variable stores `sizes_`/`strides_` in its underlying `data_` tensor, and the implementation here will not work for a Variable. Nor do we have a need to call this function on a Variable either, because we are only interested in shallow-copying a non-Variable tensor in `make_variable(...)`.

File: c10/core/TensorImpl.h
Comment (yf225) [inline c10/core/TensorImpl.h]: Same as above.

File: torch/_autograd_internal.py
Comment (yf225) [inline torch/_autograd_internal.py]: @gchanan I moved the `no_version_update()` guard to `torch._autograd_internal`, to signify that it's an internal API and should not be used by end users.

File: aten/src/ATen/SparseTensorImpl.h
Comment (gchanan) [reply aten/src/ATen/SparseTensorImpl.h]: I'm confused by a couple of things:
1) what does a "missing check" mean?
2) How would users call this on a Variable?  This is a method on an Impl -- do you mean Variable::Impl?  I don't even get how you would possibly get here, wouldn't Variable (Impl?) give you the base version, which is not a SparseTensor?

File: aten/src/ATen/function_wrapper.py
Comment (gchanan) [inline aten/src/ATen/function_wrapper.py]: is NATIVE_DYNAMIC_TYPE safe with BoolTensor, etc.?

File: test/test_cuda.py
Comment (gchanan) [reply test/test_cuda.py]: I believe we assume that TH functions won't be called directly, but only in other functions that dispatch them.  Does this mean we can avoid the bumps in TH functions?

File: torch/csrc/autograd/init.cpp
Comment (gchanan) [inline torch/csrc/autograd/init.cpp]: if these are internal, should they be prefixed with "_" or internal or something?

File: aten/src/ATen/function_wrapper.py
Comment (yf225) [reply aten/src/ATen/function_wrapper.py]: Yes it is safe with BoolTensor / IndexTensor / IntegerTensor. Examples:
```python
# BoolTensor:
{
 'api_name': '_th_lt_out',
 'formals_list': [
    {
        'dynamic_type': 'BoolTensor',
        'name': 'result',
        'output': True,
        'type': 'Tensor &'  # formal['type'] of a BoolTensor is `Tensor &`, which is what we expected
    },
    ...]
}

# IndexTensor
{
 'api_name': '_th_nonzero_out',
 'formals_list': [
    {
        'dynamic_type': 'IndexTensor',
        'name': 'result',
        'output': True,
        'type': 'Tensor &'  # formal['type'] of an IndexTensor is `Tensor &`, which is what we expected
   },
   …],
}

# IntegerTensor
{
 'api_name': '_th_pstrf_out',
 'formals_list': [...,
    {
        'dynamic_type': 'IntegerTensor',
        'name': 'res2',
        'output': True,
        'type': 'Tensor &'  # formal['type'] of an IntegerTensor is `Tensor &`, which is what we expected
   },
   …],
}

```

File: aten/src/ATen/SparseTensorImpl.h
Comment (yf225) [reply aten/src/ATen/SparseTensorImpl.h]: Yes by Variable I meant Variable::Impl. This is mostly to prevent people from accidentally doing `sparse_variable.unsafeGetTensorImpl()->shallow_copy_and_detach()` (which shallow-copies the Variable::Impl and is wrong) instead of `sparse_variable.data().unsafeGetTensorImpl()-> shallow_copy_and_detach()` (which shallow-copies the Variable::Impl's underlying `data_` tensor and is correct). We can remove this check if we feel that such misuse won't happen and we are disciplined about what we call `shallow_copy_and_detach()` on.

Comment (gchanan) [reply aten/src/ATen/SparseTensorImpl.h]: I still don't get how this would be triggered.  If we are trying to prevent `sparse_variable.unsafeGetTensorImpl()->shallow_copy_and_detach()`, but that doesn't call this, how does the assert help?  Shouldn't this be in the base TensorImpl or as an override in Variable::Impl?

File: test/test_nn.py
Comment (gchanan) [inline test/test_nn.py]: what's going on here?

File: torch/csrc/autograd/VariableTypeManual.cpp
Comment (gchanan) [inline torch/csrc/autograd/VariableTypeManual.cpp]: can you enclose this in `{ }`

**Comment by gchanan** [comment]
summary of in-person discussion: is there an existing, non-contrived (i.e., doesn't toggle VariableTypeMode manually) example that shows why we need the changes to TypeDefault.  I.e. C++ API code that works today but would break if we don't make the change?

(I'm not against the change, I want to understand if this is to support existing use cases or new use cases).

File: aten/src/ATen/SparseTensorImpl.h
Comment (yf225) [reply aten/src/ATen/SparseTensorImpl.h]: `sparse_variable.unsafeGetTensorImpl()->shallow_copy_and_detach()` does call this, because this function overrides the one in the base TensorImpl and thus the base TensorImpl one is not called when we call `sparse_variable.unsafeGetTensorImpl()->shallow_copy_and_detach()`. Variable::Impl doesn't override this because shallow-copying a TensorImpl itself is only meaningful for pure TensorImpl, not Variable::Impl.

**Comment by yf225** [comment]
Update from in-person discussion: version counter should be a concept of autograd and managed by autograd-related code. If the user is not using autograd or chooses to in-place update a variable in the non-variable scope, the version counter of that variable should not be incremented.

File: aten/src/ATen/SparseTensorImpl.h
Comment (gchanan) [inline aten/src/ATen/SparseTensorImpl.h]: could you add this comment for OpaqueTensorImpl as well?

File: c10/core/TensorImpl.h
Comment (gchanan) [inline c10/core/TensorImpl.h]: This doesn't describe *why* TensorImpl isn't on AutogradMeta.

File: aten/src/ATen/SparseTensorImpl.h
Comment (gchanan) [reply aten/src/ATen/SparseTensorImpl.h]: This explanation is incorrect.  I encourage you to try it for yourself and see.

Comment (yf225) [reply aten/src/ATen/SparseTensorImpl.h]: Yes sorry it was my mistake - `sparse_variable.unsafeGetTensorImpl()->shallow_copy_and_detach()`  calls the base TensorImpl's `shallow_copy_and_detach()`, and as long as we have the `AT_ASSERT(!is_variable())` check in the base TensorImpl we are in good shape.

File: c10/core/TensorImpl.h
Comment (yf225) [reply c10/core/TensorImpl.h]: Fixed

File: c10/core/TensorImpl.h
Comment (gchanan) [inline c10/core/TensorImpl.h]: given this explanation, a logical alternative could be creating the version counter when it's saved for a backward.  Why was the above chosen instead?

Comment (yf225) [reply c10/core/TensorImpl.h]: Improved comment to address the concern.

**Review by gchanan — APPROVED** [review]

File: c10/core/TensorImpl.h
Comment (ezyang) [inline c10/core/TensorImpl.h]: This note will be difficult for future readers to understand. The reason it is difficult to understand is it is making a comment relative to a change, but in the future, no one will see the change, they will see the code as is! Sometimes knowing the historical context is helpful to know why code is setup this way, but in this case, the issue should be explained from first principles.

File: c10/core/TensorImpl.h
Comment (ezyang) [inline c10/core/TensorImpl.h]: This has nothing to do with "forward pass" specifically. We support multithreaded read access to Tensor, period. Saving a tensor is a read operation, and therefore the reasoning below follows.

Comment (yf225) [reply c10/core/TensorImpl.h]: I feel that we could ask "why saving a tensor has to be a read operation", but it can't be a write operation because then the forward pass will not be thread-safe, and having forward pass work in multi-thread scenario is an important use case.

File: c10/core/TensorImpl.h
Comment (yf225) [reply c10/core/TensorImpl.h]: I will remove the "After the Variable/Tensor merge" phrase after the merge is completed, change the future tense to present tense and rework the comment to make it easier to understand.

File: c10/core/TensorImpl.h
Comment (ezyang) [inline c10/core/TensorImpl.h]: Do version counters update on shape change? Are you referring to `resize_`? Because if you in-place restride a tensor, I actually wouldn't expect the version counter to update (after all, the data didn't change.)

File: c10/core/TensorImpl.h
Comment (ezyang) [inline c10/core/TensorImpl.h]: I don't think it's just typical: the *only* way we should be bumping the version counter, is if we do an in-place mutation.

File: c10/core/TensorImpl.h
Comment (ezyang) [inline c10/core/TensorImpl.h]: You start a list here but actually there is only one member of the list ;)

File: c10/core/TensorImpl.h
Comment (ezyang) [inline c10/core/TensorImpl.h]: It's important to distinguish `x.detach()` from `x.data`. The former shares version counter; the latter doesn't, am I right? You should actually write code here. A link to #5396 would make it even better!

File: c10/core/TensorImpl.h
Comment (ezyang) [inline c10/core/TensorImpl.h]: No, I don't think this is accurate. The most well known escape hatch is `x.data` and the fact that version counters don't get "incremented" here has nothing to do with whether or not we increment version counter in non-Variable operations; it's the fact that we didn't share the version counter in this situation.

Comment (ezyang) [reply c10/core/TensorImpl.h]: Also, this question comes totally out of the blue. Why would I ask a question like this? It would only make sense if I was previously told, "Hey! Version counters get updated when you call in-place operations on variables. They do NOT get updated when you call in-place operations on tensors."

Comment (ezyang) [reply c10/core/TensorImpl.h]: IMO: The accurate answer to this question is that we think of "version counter tracking" as an affordance that it is given to you from the Variable API. So if you don't use the Variable API, you don't get this affordance.

File: c10/core/TensorImpl.h
Comment (ezyang) [inline c10/core/TensorImpl.h]: I'd make it clearer here that you are talking about a *hypothetical* alternative way to work around this problem, but not what is actually implemented.

File: c10/core/TensorImpl.h
Comment (ezyang) [reply c10/core/TensorImpl.h]: To me, it's obvious that saving a variable for later is a read-only operation. (Forget about PyTorch variables. If I stash an object somewhere so I can look at it later, surely that "stashing" process doesn't write to the object!)

I think it's perfectly fine to give an example where making "tensor saving" a write operation would break things. But you're more apt to confuse readers if you make this front and center.

File: aten/src/ATen/SparseTensorImpl.h
Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.h]: I'm not a big fan of copy pasting... it would be better to have a `See Note [blah blah]` cross-reference instead

File: c10/core/TensorImpl.h
Comment (yf225) [reply c10/core/TensorImpl.h]: `VariableType::as_strided_(self, size, stride, storage_offset)` actually updates the version counter of `self`. We can investigate whether this is strictly necessary, outside of this PR.

Comment (yf225) [reply c10/core/TensorImpl.h]: I looked at all functions in `VariableType` that does in-place restride to a tensor (`VariableType::set_`, `VariableType::_th_set_`, `VariableType::as_strided_`), and all of them resize the tensor at the same time, which is why the version counter is bumped in those functions.

In principle we shouldn't update the version counter if we only in-place restride a tensor. I will change "shape" to "size" in this comment.

Comment (gchanan) [reply c10/core/TensorImpl.h]: there's no API for just restriding, so this doesn't seem relevant.  But if there was an API that let you raw inplace restride a Tensor I would absolutely expect it to update the version counter because it can change the data.

**Pull request closed without merging**
